Simply `GNUmakefile` and add support for ARM64

walro opened the pull request:
See individual commits. Just the simplification of the `GNUmakefile` would do the trick, but I deemed it to be some value in keeping the commit message of the fist commit.

---

Investigation of the different values of `uname` for different combinations (not really relevant after simplification of makefile, but consider it fun facts):

| system             | `uname -m` | `uname -s` | `uname -p` |
|--------------------|------------|------------|------------|
| macOS 13.1 (M1) | arm64 | Darwin | arm |
| macOS 13.1 (Intel) | x86_64 | Darwin | i386 |
| Debian 11 (arm - container) | aarch64 | Linux | unknown |
| Ubuntu 18.04 (Intel) | x86_64 | Linux | x86_64 |
| Ubuntu 20.04 (Intel) | x86_64 | Linux | x86_64 |
| Debian 11 via Docker on macOS 13.1 (M1) with `--platform linux/arm64` | aarch64 | Linux | unknown |
| Debian 11 via Docker on macOS 13.1 (M1) with `--platform linux/amd64` | x86_64 | Linux | unknown |
| Debian 11 via WSL2 (Windows 11 in VMware on macOS 12.6.2 (Intel)) | x86_64 | Linux | unknown |
| Ubuntu 20.04 via Docker on macOS 13.1 (M1) with `--platform linux/arm64` | aarch64 | Linux | aarch64 |
| Ubuntu 20.04 via Docker on macOS 13.1 (M1) with `--platform linux/amd64` | x86_64 | Linux | x86_64 |

walro commented on GNUmakefile:
```diff
@@ -24,7 +24,9 @@ ifeq ($(UNAME_P),i386)
 	endif
 else
 	ifeq ($(UNAME_P),AMD64)
```
Think we saw that `uname -p` could give `unknown` (or something along those lines) under WSL: :)

dentarg replied:
Got myself a VM in Fusion Player with Windows 11 Pro N and WSL 2. In Ubuntu (20.04 and 22.04) `uname -p` reports `x86_64` but in Debian 11 it reports `unknown`.

walro replied:
Added a table in the PR description we can add info to

walro replied:
`go tool dist list` shows all supported platforms (which I believe is the combination of `GOOS` + `GOARCH`):

```
$ go tool dist list

aix/ppc64
android/386
android/amd64
android/arm
android/arm64
darwin/amd64
darwin/arm64
dragonfly/amd64
freebsd/386
freebsd/amd64
freebsd/arm
freebsd/arm64
illumos/amd64
ios/amd64
ios/arm64
js/wasm
linux/386
linux/amd64
linux/arm
linux/arm64
linux/loong64
linux/mips
linux/mips64
linux/mips64le
linux/mipsle
linux/ppc64
linux/ppc64le
linux/riscv64
linux/s390x
netbsd/386
netbsd/amd64
netbsd/arm
netbsd/arm64
openbsd/386
openbsd/amd64
openbsd/arm
openbsd/arm64
openbsd/mips64
plan9/386
plan9/amd64
plan9/arm
solaris/amd64
windows/386
windows/amd64
windows/arm
windows/arm64
```

So for instance there seems to be no support for 32 bit at all on macOS.

walro committed:
Add support for building for ARM64

... and drop (not working support, see below) support for 32-bit. I don't think anyone will miss it.

I don't think that 32-bit ever worked, because `i386` is not valid a valid value for `GOARCH`:

```
$ GOOS=linux GOARCH=i386 go build -ldflags "-X 'main.version=0.0.1'" -o terraform-provider-elephantsql_v0.0.1
go: unsupported GOOS/GOARCH pair linux/i386
```

It should be just `386`.

walro commented:
Only tested on Mac (ARM and Intel mode), guess we should verify on linux under x64/aarch64.

To get rudimentary 32-bit support we could let it fall through to that in an else branch, but not sure we should bother 🤷.

walro commented on GNUmakefile:
```diff
@@ -4,28 +4,20 @@ PROVIDER_VERSION = 0.0.1
-
 UNAME_S := $(shell uname -s)
 ifeq ($(UNAME_S),Linux)
 	GOOS += linux
```
Hm, why do we do `+= ` here, it doesn't really work that well if `$GOOS` is set (same for `GOARCH`).

Also, I wonder why we even do this dance at all. If we only want to ensure that we create a valid `PROVIDER_ARCH` I think we can use `go env` instead to get "current state":

```shell
$ go env GOOS
darwin
$ go env GOARCH
arm64
```

Maybe I oversee something.

walro replied:
So what I am suggesting is this:

```diff
diff --git a/GNUmakefile b/GNUmakefile
index b6b4fe4..973f0e3 100644
--- a/GNUmakefile
+++ b/GNUmakefile
@@ -4,21 +4,8 @@ PROVIDER_VERSION = 0.0.1

 default: build

-UNAME_S := $(shell uname -s)
-ifeq ($(UNAME_S),Linux)
-       GOOS += linux
-else ifeq ($(UNAME_S),Darwin)
-       GOOS += darwin
-endif
-
-UNAME_M := $(shell uname -m)
-ifeq ($(UNAME_M),x86_64)
-       GOARCH += amd64
-else ifeq ($(UNAME_M),arm64)
-       GOARCH += arm64
-else ifeq ($(UNAME_M),aarch64)
-       GOARCH += arm64
-endif
+GOOS = $(shell go env GOOS)
+GOARCH = $(shell go env GOARCH)
 PROVIDER_ARCH = $(GOOS)_$(GOARCH)

 tools:
@@ -32,9 +19,7 @@ local-clean:
        rm -rf ~/.terraform.d/plugins/localhost/elephantsql/elephantsql/$(PROVIDER_VERSION)/$(PROVIDER_ARCH)/terraform-provider-elephantsql_v$(PROVIDER_VERSION)

 local-build: local-clean
-       @echo $(GOOS);
-       @echo $(GOARCH);
-       GOOS=$(GOOS) GOARCH=$(GOARCH) go build -ldflags "-X 'main.version=$(PROVIDER_VERSION)'" -o terraform-provider-elephantsql_v$(PROVIDER_VERSION)
+       go build -ldflags "-X 'main.version=$(PROVIDER_VERSION)'" -o terraform-provider-elephantsql_v$(PROVIDER_VERSION)

 local-install: local-build
        mkdir -p ~/.terraform.d/plugins/localhost/elephantsql/elephantsql/$(PROVIDER_VERSION)/$(PROVIDER_ARCH)
```

This makes it possible to build for other architectures too, like: `GOARCH=amd64 make local-install`.

dentarg replied:
Looks good to me, commit?

walro replied:
Pushed in [52e7df4](https://github.com/ElephantSQL/terraform-provider-elephantsql/pull/31/commits/52e7df4cbab9f59b103171c83160868c0faa763a), intentionally did not force push as I wanted to preserve the earlier commit.

walro committed:
Simplify `GNUmakefile`

Instead of relying on what `uname` states for different machines, let `go` tell us.

This also has the upshot of making it possible to build for any architecture by doing, for instance: `GOARCH=amd64 make local-install`

dentarg reviewed (approved):
Nice work, I can build in WSL2 (Debian). Tested the provider in macOS too.
